src/orchestrator.ts

```diff
@@ -550,28 +542,28 @@ export class Orchestrator {
         }
     }
 
-    private all(state: HistoryEvent[], tasks: Task[]): TaskSet {
-        const faulted = tasks.filter(TaskFilter.isFailedTask);
-        if (faulted.length > 0) {
-            return TaskFactory.FailedTaskSet(tasks, (faulted[0] as FailedSingleTask).exception);
-        }
-
-        const succesfulTasks = tasks.filter(TaskFilter.isSuccessfulSingleTask);
-        const isCompleted = succesfulTasks.length === tasks.length;
-        if (isCompleted) {
-            const results = succesfulTasks.reduce((acc, t) => {
-                return [...acc, t.result];
-            }, []);
+    private all(state: HistoryEvent[], tasks: TaskBase[]): TaskSet {
+        const completedTasks = tasks.filter(TaskFilter.isCompletedTask);
+        if (completedTasks.length === tasks.length) {
+            const sortedTasks = completedTasks.sort(TaskFilter.CompareFinishedTime);
+            const completionIndex = sortedTasks[tasks.length - 1].completionIndex;
```

Comment:
Is it okay that the resulting `TaskSet` has a `completionIndex` that is the same as one of the completed `TaskBase` objects? I would have expected there to be a `+ 1` somewhere in here. Does the relative order of `TaskBase` and `TaskSet` not matter?

Reply:
A TaskSet returned by `all()` is considered finished when its last subtask finishes, so I think it is fair that it would have the same completion index as that subtask.

Imagine the following C#

```csharp

// Tasks will finish in order task1 then task2
Task firstTask =  task1();
Task secondTask  = task2();

Task bothFinishedTask= Task.WhenAll(firstTask, secondTask);

Task lastTask = Task.WhenAny(bothFinishedTask, secondTask);
```

In C#, technically we would probably expect that `lastTask == secondTask`, just due to the latency it takes for `bothFinishedTask` to realize that it is finished after `secondTask` finishes.

With my current implementation, I think I would expect that `lastTask == bothFinishedTask`, just due to the ordering of parameters in `Task.WhenAny()`. I don't think this is a problem, as long as it is deterministic.

In general, I wouldn't expect the above scenario very often regardless. 

If that is not acceptable, I could see doing a ` + .5`, just to make sure that it doesn't interfere with any other task's completion indices.